File: C++/GPT-4/Dynamic Programming/Hard/Solution/818. Race Car/2.cpp
class Solution {
public:
    int racecar(int target) {
        vector<int> dp(target + 2);
        for (int i = 1; i <= target; i++) {
            dp[i] = INT_MAX;
            int m = 1, j = 0;
            for (; m < i; m = (1 << ++j) - 1)
                for (int q = 0, p = 0; p < m; p = (1 << ++q) - 1)
                    dp[i] = min(dp[i], j + 1 + q + 1 + dp[i - (m - p)]);
            dp[i] = min(dp[i], j + (i == m ? 0 : 1 + dp[m - i]));
        }
        return dp[target];
    }
};

File: Java/GPT-4/Queue/Hard/Solution/862. Shortest Subarray with Sum at Least K/2.java
import java.util.*;

class Solution {
    public int shortestSubarray(int[] A, int K) {
        int N = A.length;
        long[] P = new long[N+1];
        for (int i = 0; i < N; ++i)
            P[i+1] = P[i] + (long) A[i];

        int ans = N+1; //sentinel value
        Deque<Integer> monoq = new LinkedList<>(); //store index of "good" elements of P in increasing order

        for (int y = 0; y < P.length; ++y) {
            //Want opt(y) = largest x with P[x] <= P[y] - K;
            while (!monoq.isEmpty() && P[y] <-= P[monoq.getLast()])
                monoq.removeLast();
            while (!monoq.isEmpty() && P[y] >= P[monoq.getFirst()] + K)
                ans = Math.min(ans, y - monoq.removeFirst());

            monoq.addLast(y);
        }

        return ans < N+1 ? ans : -1;
    }
}

